designdocs/user-handles.md

```diff
@@ -0,0 +1,308 @@
+
+## Sequence Diagram
+
+```mermaid
+sequenceDiagram
+    participant Wallet
+    actor User
+    participant App
+    participant RPC as RPC Node
+    participant Chain as Frequency
+    User->>App: Enter desired handle
+    App->>RPC: get_suffix_options(handle)
+    RPC->>Chain: Query current seed
+    Chain-->>RPC: Return seed
+    RPC->>RPC: Compute suffix options
+    RPC->>Chain: check_handle_availability(options)
+    Chain-->>RPC: Return subset of options
+    RPC-->>App: Return subset of options
+    App->>User: Present suffix options
+    User->>App: Choose desired suffix
+    App->>Wallet: Send signup data
+    Wallet->>Wallet: Perform 2FA if first attempt
+    Wallet-->>App: Return signed payload
+    App->>RPC: create_msa_handle(..., handle, suffix)*
+    RPC->>Chain: Submit transaction
+    Chain->>Chain: Check validity and availability
+    Chain-->>App: MsaHandleCreated event with handle
+    App-->>RPC: Publish Profile Announcement/batch with (canonical) DSNP handle
+    App->>User: Proceed with setup
+    User->>App: Request to get MSA ID for handle
+    App->>RPC: get_msa_id_for_handle(handle)
+    RPC->>Chain: Query handle mapping state (State Query MSA<->Handle)
+    Chain-->>RPC: Return handle mapping state (State Query MSA<->Handle)
+    RPC-->>App: Return MSA ID for handle
+    User->>App: Request to get handle for MSA ID
+    App->>RPC: get_handle_for_msa_id(msa_id)
+    RPC->>Chain: Query MSA ID mapping state (Reverse Index Handle<->MSA)
+    Chain-->>RPC: Return MSA ID mapping state (Reverse Index Handle<->MSA)
+    RPC-->>App: Return handle for MSA ID
+```
+
+## Storage
+
+* **HandleSuffixMin**: This storage value (```u32``) will keep track of the minimum suffix value allowed for a given handle.
+* **HandleSuffixMax**: This storage value (```u32``) will keep track of the maximum suffix value allowed for a given handle.
+* **HandleRetirementPeriod**: This storage value (```u32``) will keep track of the number of blocks a handle must be retired before it can be reused.
+* **MSAIdToUserHandle**: This storage will map MSAIds to their corresponding user handles.
+* **UserHandlesToMSAId**: This storage will map user handles to their corresponding MSAIds.
+* **RetiredHandles**: This storage will keep track of all retired handles and the block number at which they were retired. This probably is a spike to run few ideas on how to implement this.
+* **UsedSuffixes**: This storage will keep track of all used suffixes to ensure that no two handles have the same suffix. (May not be necessary)
+* **Seed (current and previous)**: This storage values (```tuple (s0, s1)```) will keep track of the current and previous seed values.
+
+## Required Extrinsics
+
+### Primitives
+
+``` rust
+
+MsaHandlePayload {
+    handle: &[u8],
+    suffix: u32,
+    msa_id: u64,
```

Comment:
We can derive the MSA Id from the signing key. (You can only create a handle for your own MSA, and keys can only be attached to one MSA at a time)